ViewEnvironmentUI/Sources/ViewEnvironmentPropagatingObject.swift

```diff
@@ -0,0 +1,272 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import ViewEnvironment
+
+/// A protocol describing a ``ViewEnvironmentPropagating`` object that can:
+/// - Override the ancestor and descendants
+/// - Add environment update observations
+/// - Flag the backing object for needing to have the `ViewEnvironment` reapplied (e.g. `setNeedsLayout()`)
+///
+/// This protocol was abstracted to share propagation logic between `UIViewController` and `UIView`'s support for
+/// ``ViewEnvironmentPropagating``, but could be used for any object-based node that wants to support
+/// `ViewEnvironment` propagation.
+///
+public protocol ViewEnvironmentPropagatingObject: AnyObject, ViewEnvironmentPropagating {
+    /// The default ancestor for `ViewEnvironment` propagation.
+    ///
+    @_spi(ViewEnvironmentWiring)
+    var defaultEnvironmentAncestor: ViewEnvironmentPropagating? { get }
+
+    /// The default descendants for `ViewEnvironment` propagation.
+    ///
+    @_spi(ViewEnvironmentWiring)
+    var defaultEnvironmentDescendants: [ViewEnvironmentPropagating] { get }
+
+    /// Informs the backing object that this specific node should be flagged for another application of the
+    /// `ViewEnvironment`.
+    ///
+    /// For `UIViewController`/`UIView`s this typically corresponds to `setNeedsLayout()`.
+    ///
+    @_spi(ViewEnvironmentWiring)
+    func setNeedsApplyEnvironment()
+}
+
+extension ViewEnvironmentPropagating where Self: ViewEnvironmentPropagatingObject {
+    /// Consumers _must_ call this function when the environment should be re-applied, e.g. in
+    /// `viewWillLayoutSubviews()` for `UIViewController`s and `layoutSubviews()` for `UIView`s.
+    ///
+    /// This will call ``apply(environment:)`` on the receiver if the node has been flagged for needing update.
+    ///
+    public func applyEnvironmentIfNeeded() {
+        guard needsEnvironmentUpdate else { return }
+
+        needsEnvironmentUpdate = false
+
+        if let observing = self as? ViewEnvironmentObserving {
+            let environment = observing.environment
+            observing.apply(environment: environment)
+        }
+    }
+}
+
+extension ViewEnvironmentPropagatingObject {
+    /// Adds a `ViewEnvironment` change observation.
+    ///
+    /// The observation will only be active for as long as the returned lifetime is retained or
+    /// `cancel()` is called on it.
+    ///
+    @_spi(ViewEnvironmentWiring)
+    public func addEnvironmentNeedsUpdateObserver(
+        _ onNeedsUpdate: @escaping (ViewEnvironment) -> Void
+    ) -> ViewEnvironmentUpdateObservationLifetime {
+        let object = NSObject()
```

Comment:
Thought: Is it worth making a special type / subclass, so the memory debugger, etc, shows the type more obviously in case something goes wrong?